TorchSharp/TypeGeneration.tt

```diff
@@ -541,21 +539,51 @@ foreach (var type in new [] {
         extern static HType TH<#=tname#>Tensor_newWithStorage1d(<#=tname#>Storage.HType handle, UIntPtr offset, long size, long stride);
 
         /// <summary>
-        ///   Access to element at the specified position in the tensor
+        ///   Create a new tensor with a specified storage arena.
+        /// </summary>        
+        /// <param name="storage">Storage arena.</param>
+        /// <param name="offset">Offset within the input storage the storage of the new tensor will start from.</param> 
+        /// <param name="size">Size of the first dimension.</param>     
+        /// <param name="stride">Stride of the first dimension.</param>          
+        public static <#=tname#>Tensor NewWithStorage1d(<#=tname#>Tensor.<#=tname#>Storage storage, UIntPtr offset, long size, long stride)
+        {
+            var result = new <#=tname#>Tensor(TH<#=tname#>Tensor_newWithStorage1d(storage.handle, offset, size, stride));
+            result.storage = storage;
```

Comment:
Not sure about this. Can you please make sure that `result.storage.handle == storage.handle`. Because if they are different (which I think they are) we are making some assumptions here that may create bugs.

Reply:
This comment applies for the following functions too.

Reply:
I see what you mean but I don't understand the sentence "having managed state means that in the future, when APIs surface a type, we have no way of retrieving "attached" data (like this field in Storage)". As it is now storage contains no data, it is just a memoization for the call to `TH<#=tname#>Tensor_storage (handle)`. I added this because we were creating (what I thought were) unnecessary objects.

Reply:
This is a very common scenario when doing native bindings like this, and while it is solvable, it introduces a tremendous amount of complexity of dealing with correctly.   It happens in almost every binding where you have pure C# objects that are proxies to unmanaged objects.

The bug happens like this.

You create a managed object that is a peer to a native object, with additional data, then you pass the instance of this to some unmanaged object, for example "fooBar.SetTensor (t)", now you let "t" go out of scope and the reference managed by managed code drops and the object is collected.   Now, at some point in the future, the type is resurfaced, for example "fooBar.GetTensor()" call.   

When you have a pure peer, it is cheap to create those objects, you merely create a new wrapper type by setting the handle, and you are off to the races.   But if you had state, the state is now gone.

Now, it might be possible that we can always recreate this particular one, but if this was out of sync, then we would have a situation where those do not match.

To solve this problem, you now need to keep a GCHandle to the object, to ensure that as long as the unmanaged reference exists, the managed object is kept alive.   This complicates things, because now only unmanaged code can properly release the data, as the GC would not be able to do its job (as far as the GC is concerned, the managed object is retained by the GCHandle).

So now we need to deal with some sort of notification system for when the unmanaged object is released.   Some frameworks have it, some do not.    This one does not.

Now, it is possible that this is not a problem for storage, but I would rather not have any managed state in the pure peers that forces us to come up to a solution for the above problem.

Btw, the above points out that we probably should seal these types as well.

Reply:
I have removed the memoization logic.